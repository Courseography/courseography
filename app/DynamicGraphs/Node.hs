module DynamicGraphs.Node (lookupCourse) where

-- TODO: Rename this file
import qualified Data.Text.Lazy as T
import Database.CourseQueries (prereqsForCourse)
import Database.Requirement (Req(..))
import WebParsing.ReqParser (parseReqs)

lookupCourse :: T.Text -> IO [(T.Text, Req)]
lookupCourse code = do
    prereqResults <- prereqsForCourse $ T.toStrict code
    case prereqResults of
        Left _ -> return []
        Right prereqStr ->
            do
            let prereqs = parseReqs (T.unpack $ T.fromStrict prereqStr)
            rest <- lookupReqs prereqs
            return $ (code, prereqs) : rest

lookupReqs :: Req -> IO [(T.Text, Req)]
lookupReqs (J name _) = lookupCourse $ T.pack name
lookupReqs (AND parents) = do
    parentReqs <- mapM lookupReqs parents
    return (concat parentReqs)
lookupReqs (OR parents) = do
    parentReqs <- mapM lookupReqs parents
    return (concat parentReqs)
lookupReqs (FCES _ parent) = lookupReqs parent
lookupReqs (GRADE _ parent) = lookupReqs parent
-- This will catch both NONE and RAW values.
lookupReqs _ = return []
