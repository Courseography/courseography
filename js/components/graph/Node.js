import PropTypes from "prop-types"
import React from "react"
import { refLookUp } from "../common/utils"

/** React component class representing a Node on the graph
 *
 * States:
 *  - status: holds the current status message (see below)
 *  - selected: whether the node is selected, based on the status message
 *
 *  Unselected status messages:
 *    - takeable: all prerequisites are satisfied (or no prereqs/parents)
 *    - inactive: missing some prerequisites
 *
 *  Selected status messages:
 *    - active: all prerequisities are satisfied
 *    - overridden: missing some prerequisities (will have a red border)
 *
 *  On Hover status message:
 *    - missing: means that this node is a prerequisite node that is not satisfied  (red border)
 *
 * Types of nodes:
 *  - Course nodes are nodes that represent a certain course
 *  - Hybrid nodes are the smaller, grey nodes on the graph that represent another course node
 *    farther away. They can only be either 'active' or 'inactive'
 */
export default class Node extends React.Component {
  /** Create a node */
  constructor(props) {
    super(props)
    var state = localStorage.getItem(this.props.JSON.id_)
    if (this.props.editMode) {
      state = "" // TODO: define what editMode is
    } else if (state === null) {
      state = this.props.parents.length === 0 ? "takeable" : "inactive"
    }
    this.state = {
      status: state,
      selected: ["active", "overridden"].indexOf(state) >= 0,
    }
  }

  /**
   * Checks whether this Node is selected
   * @return {boolean}
   */
  isSelected = () => {
    if (this.props.hybrid) {
      return this.state.status === "active"
    } else {
      return this.state.selected
    }
  }

  /**
   * Checks whether all prerequisite/preceding nodes for the current one are satisfied
   * @return {boolean}
   */
  arePrereqsSatisfied = () => {
    var svg = this.props.svg
    /**
     * Recursively checks that preceding nodes are selected
     * @param  {string|Array} element Node(s)/other on the graph
     * @return {boolean}
     */
    function isAllTrue(element) {
      if (typeof element === "string") {
        if (svg.nodes.current[element] !== undefined) {
          return svg.nodes.current[element].isSelected()
        } else if (svg.bools.current[element] !== undefined) {
          return svg.bools.current[element].isSelected()
        } else {
          return false
        }
      } else {
        return element.some(isAllTrue)
      }
    }

    return this.props.parents.every(isAllTrue)
  }

  /**
   * Update the state/status of a node (and its children/edges)
   * @param  {boolean} recursive whether we should recurse on its children
   */
  updateNode = recursive => {
    var newState
    if (this.arePrereqsSatisfied()) {
      if (this.isSelected() || this.props.hybrid) {
        newState = "active"
      } else {
        newState = "takeable"
      }
    } else {
      if (this.isSelected() && !this.props.hybrid) {
        newState = "overridden"
      } else {
        newState = "inactive"
      }
    }

    var nodeId = this.props.JSON.id_

    // Updating the children will be unnecessary if the selected state of the current node has not
    // changed, and the original state was not 'missing'
    if (
      ["active", "overridden"].indexOf(newState) >= 0 ===
        ["active", "overridden"].indexOf(this.state.status) >= 0 &&
      this.state.status !== "missing"
    ) {
      localStorage.setItem(nodeId, newState)
      this.setState({ status: newState })
      return
    }

    if (recursive === undefined || recursive) {
      var svg = this.props.svg
      this.setState({ status: newState }, function () {
        localStorage.setItem(nodeId, newState)
        this.props.childs.forEach(function (node) {
          var currentNode = refLookUp(node, svg)
          if (currentNode !== undefined) {
            currentNode.updateNode()
          }
        })
        var allEdges = this.props.outEdges.concat(this.props.inEdges)
        allEdges.forEach(edge => {
          var currentEdge = svg.edges.current[edge]
          if (currentEdge !== undefined) {
            currentEdge.updateStatus()
          }
        })
      })
    } else {
      this.setState({ status: newState })
      localStorage.setItem(nodeId, newState)
    }
  }

  /** Controls the selection and deselection of a node by switching states and updating the graph */
  toggleSelection = () => {
    this.setState({ selected: !this.state.selected }, function () {
      this.updateNode()
    })
  }

  /** Sets the status of all missing prerequisites to 'missing' */
  focusPrereqs = () => {
    var svg = this.props.svg
    // Missing prerequisites need to have their status updated to 'missing'
    if (["inactive", "overridden", "takeable"].indexOf(this.state.status) >= 0) {
      this.setState({ status: "missing" }, () => {
        this.props.inEdges.forEach(edge => {
          var currentEdge = svg.edges.current[edge]
          if (currentEdge === null || currentEdge === undefined) {
            return
          }
          var sourceNode = refLookUp(currentEdge.props.source, svg)
          if (!sourceNode.isSelected()) {
            currentEdge.setState({ status: "missing" })
          }
        })
        this.props.parents.forEach(node => {
          if (typeof node === "string") {
            var currentNode = refLookUp(node, svg)
            if (currentNode !== undefined) {
              currentNode.focusPrereqs()
            }
          } else {
            node.forEach(n => {
              var currentNode = refLookUp(n, svg)
              if (currentNode !== undefined) {
                currentNode.focusPrereqs()
              }
            })
          }
        })
      })
    }
  }

  /**
   * Resets 'missing' nodes and edges to the previous statuses:
   *  active, inactive, overridden, takeable
   */
  unfocusPrereqs = () => {
    var svg = this.props.svg
    this.updateNode(false)
    this.props.parents.forEach(function (node) {
      if (typeof node === "string") {
        var currentNode = refLookUp(node, svg)
        currentNode.unfocusPrereqs()
      } else {
        node.forEach(n => {
          var currentNode = refLookUp(n, svg)
          currentNode.unfocusPrereqs()
        })
      }
    })
    this.props.inEdges.forEach(function (edge) {
      var currentEdge = svg.edges.current[edge]
      if (currentEdge.state.status === "missing") {
        currentEdge.updateStatus()
      }
    })
  }

  getDataTestId = () => {
    if (this.props.hybrid) {
      return `h(${this.props.parents.join(",")})`
    }
    return this.props.JSON.id_
  }

  render() {
    let ellipse = null
    var newClassName = this.props.className + " " + this.state.status
    if (this.props.highlighted) {
      var attrs = this.props.JSON
      var width = parseFloat(attrs.width) / 2
      var height = parseFloat(attrs.height) / 2
      ellipse = (
        <ellipse
          className="spotlight"
          cx={parseFloat(attrs.pos[0]) + width}
          cy={parseFloat(attrs.pos[1]) + height}
          rx={width + 9}
          ry={height + 8.5}
        />
      )
    }

    var gAttrs = {
      textRendering: "geometricPrecision",
      shapeRendering: "geometricPrecision",
      onKeyDown: this.props.svg.onKeyDown,
      onWheel: this.props.svg.onWheel,
      onMouseEnter: this.props.onMouseEnter,
      onMouseLeave: this.props.onMouseLeave,
      onClick: this.props.onClick,
    }

    var rectAttrs = {
      height: this.props.JSON.height,
      width: this.props.JSON.width,
      x: this.props.JSON.pos[0],
      y: this.props.JSON.pos[1],
    }

    if (this.props.className === "node") {
      rectAttrs["rx"] = "8"
      rectAttrs["ry"] = "8"
    }

    var rectStyle = {
      fill: this.props.JSON.fill,
    }

    var textXOffset = this.props.JSON.pos[0] + this.props.JSON.width / 2

    // TODO: Look at this.props to see what we need to give the g
    return (
      <g
        {...gAttrs}
        id={this.props.JSON.id_}
        className={newClassName}
        data-testid={this.getDataTestId()}
      >
        {ellipse}
        <rect
          {...rectAttrs}
          style={rectStyle}
          filter={
            this.props.className === "hybrid"
              ? ""
              : `url(#${this.props.nodeDropshadowFilter})`
          }
        />
        {this.props.JSON.text.map(function (textTag, i) {
          var textAttrs = {
            x: textXOffset,
            y: textTag.pos[1],
          }
          return (
            <text {...textAttrs} key={i}>
              {textTag.text}
            </text>
          )
        })}
      </g>
    )
  }
}

Node.propTypes = {
  childs: PropTypes.array,
  className: PropTypes.string,
  editMode: PropTypes.bool,
  highlighted: PropTypes.bool,
  hybrid: PropTypes.bool,
  inEdges: PropTypes.array,
  JSON: PropTypes.object,
  onClick: PropTypes.func,
  onMouseEnter: PropTypes.func,
  onMouseLeave: PropTypes.func,
  outEdges: PropTypes.array,
  parents: PropTypes.array,
  svg: PropTypes.object,
  nodeDropshadowFilter: PropTypes.string,
}
