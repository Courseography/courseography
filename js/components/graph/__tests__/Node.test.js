import React from "react"
import { shallow } from "enzyme"
import { fireEvent } from "@testing-library/react"
import Node from "../Node"
import TestGraph from "./TestGraph"

describe("Hybrid Node", () => {
  it("should match snapshot", () => {
    const hybridNodeProps = {
      JSON: {
        fill: "#888888",
        graph: 1,
        height: 24,
        id_: "h46",
        pos: [18.168848, 471.586698],
        stroke: "",
        text: [
          {
            align: "begin",
            fill: "",
            graph: 1,
            pos: [20.753048200000002, 481.389408],
            rId: "text454",
            text: "CSC318/418/",
            type_: "Hybrid",
            width: 65.207497,
          },
          {
            align: "begin",
            fill: "",
            graph: 1,
            pos: [32.698848, 493.640408],
            rId: "text456",
            text: "301/384",
            length: 2,
            type_: "Hybrid",
            width: 65.207497,
          },
        ],
        type_: "Hybrid",
        width: 65.207497,
      },
      childs: ["csc404"],
      className: "hybrid",
      hybrid: true,
      inEdges: [],
      logicalType: "AND",
      outEdges: ["p32"],
      parents: ["csc301", "csc318", "csc384", "csc418"],
      svg: {
        onKeyDown: jest.fn(),
      },
    }
    const wrapper = shallow(<Node {...hybridNodeProps} />)
    expect(wrapper).toMatchSnapshot()
  })

  it("should have the 'hybrid' CSS class", async () => {
    const graph = await TestGraph.build()
    const hybrid101 = graph.getByTestId("h(aaa101)")
    expect(hybrid101.classList.contains("hybrid")).toBe(true)
  })
  it("shouldn't do anything when you hover or click it", async () => {
    const graph = await TestGraph.build()
    const hybrid101 = graph.getByTestId("h(aaa101)")

    // convert DOMTokenList to object so we can deep copy
    const cssClassDeepCopy = JSON.parse(JSON.stringify(hybrid101.classList))

    fireEvent.click(hybrid101)
    // need to convert the classList into an object on order to use Jest's compare
    expect({ ...hybrid101.classList }).toEqual(cssClassDeepCopy)
    fireEvent.click(hybrid101)
    expect({ ...hybrid101.classList }).toEqual(cssClassDeepCopy)
    fireEvent.mouseOver(hybrid101)
    expect({ ...hybrid101.classList }).toEqual(cssClassDeepCopy)
    fireEvent.mouseOut(hybrid101)
    expect({ ...hybrid101.classList }).toEqual(cssClassDeepCopy)
  })
  it("should be 'inactive' when it's prereq parent is NOT met", async () => {
    const graph = await TestGraph.build()
    const hybrid101 = graph.getByTestId("h(aaa101)")
    expect(hybrid101.classList.contains("inactive")).toBe(true)
  })
  it("should be 'active' when its prereq parent is met", async () => {
    const graph = await TestGraph.build()
    const hybrid101 = graph.getByTestId("h(aaa101)")
    const aaa101 = graph.getByTestId("aaa101")
    fireEvent.click(aaa101)
    expect(hybrid101.classList.contains("active")).toBe(true)
  })

  it("should be 'missing' if not 'active' and it's an unmet prereq of the currently hovered course", async () => {
    const graph = await TestGraph.build()
    const hybrid101 = graph.getByTestId("h(aaa101)")
    const aaa303 = graph.getByTestId("aaa303")

    fireEvent.mouseOver(aaa303)
    expect(hybrid101.classList.contains("missing")).toBe(true)
  })
})

describe("Course Node", () => {
  it("should match shallow snapshot", () => {
    const courseProps = {
      JSON: {
        fill: "#5dd5b8",
        graph: 1,
        height: 30,
        id_: "csc108",
        pos: [594.294128, 38.09390799999999],
        stroke: "",
        text: [
          {
            align: "begin",
            fill: "",
            graph: 1,
            pos: [596.555968, 59.379107999999995],
            rId: "text102",
            text: "CSC108",
            type_: "Node",
            width: 65,
          },
        ],
      },
      childs: ["csc148"],
      className: "node",
      highlighted: false,
      hybrid: false,
      inEdges: [],
      onClick: jest.fn(),
      onMouseEnter: jest.fn(),
      onMouseLeave: jest.fn(),
      outEdges: ["p6"],
      parents: [],
      svg: {},
    }

    const wrapper = shallow(<Node {...courseProps} />)
    expect(wrapper).toMatchSnapshot()
  })
  it("should have the CSS class: 'node'", async () => {
    const graph = await TestGraph.build()
    const aaa100 = graph.getByTestId("aaa100")
    expect(aaa100.classList.contains("node")).toBe(true)
  })

  describe("Unselected Course Node", () => {
    it("should be 'takeable' if it has no prereqs", async () => {
      const graph = await TestGraph.build()
      const aaa101 = graph.getByTestId("aaa101")
      expect(aaa101.classList.contains("takeable")).toBe(true)
    })
    it("should be 'inactive' if the prereqs are NOT met", async () => {
      const graph = await TestGraph.build()
      const aaa201 = graph.getByTestId("aaa201")
      expect(aaa201.classList.contains("inactive")).toBe(true)
    })
    it("should be 'takeable' if the prereqs are met", async () => {
      const graph = await TestGraph.build()
      const aaa101 = graph.getByTestId("aaa101")
      const aaa201 = graph.getByTestId("aaa201")
      fireEvent.click(aaa101) // the only prereq for AAA201
      expect(aaa201.classList.contains("takeable")).toBe(true)
    })
    it('should be "missing" when unselected and hovered over', async () => {
      const graph = await TestGraph.build()
      const aaa100 = graph.getByTestId("aaa100")
      fireEvent.mouseOver(aaa100)
      expect(aaa100.classList.contains("missing")).toBe(true)
    })

    it("when hovered, should set all unmet prereqs and itself as 'missing'", async () => {
      const graph = await TestGraph.build()
      const aaa101 = graph.getByTestId("aaa101")
      const aaa201 = graph.getByTestId("aaa201")
      const aaa102 = graph.getByTestId("aaa102")
      const aaa303 = graph.getByTestId("aaa303")

      fireEvent.mouseOver(aaa303)
      expect(aaa101.classList.contains("missing")).toBe(true)
      expect(aaa102.classList.contains("missing")).toBe(true)
      expect(aaa201.classList.contains("missing")).toBe(true)
      expect(aaa303.classList.contains("missing")).toBe(true)
    })
  })

  describe("Selected Course Node", () => {
    it("with met prereqs should 'active'", async () => {
      const graph = await TestGraph.build()
      const aaa100 = graph.getByTestId("aaa100")
      fireEvent.click(aaa100)
      expect(aaa100.classList.contains("active")).toBe(true)
    })
    describe("selected course with un-met prereqs", () => {
      it("should be 'overridden' (if you don't hover over it)", async () => {
        const graph = await TestGraph.build()
        const aaa201 = graph.getByTestId("aaa201")
        fireEvent.click(aaa201)
        expect(aaa201.classList.contains("overridden")).toBe(true)
      })

      it("'overridden' course counts as a satisfied prereq (acts like 'active')", async () => {
        const graph = await TestGraph.build()
        const aaa201 = graph.getByTestId("aaa201")
        const aaa202 = graph.getByTestId("aaa202")
        fireEvent.click(aaa201)
        expect(aaa201.classList.contains("overridden")).toBe(true)
        expect(aaa202.classList.contains("takeable")).toBe(true)
      })

      it("hovering over a course with an 'overridden' prereq does not mark the 'overridden' unmet prereqs as 'missing'", async () => {
        const graph = await TestGraph.build()
        const aaa101 = graph.getByTestId("aaa101")
        const aaa201 = graph.getByTestId("aaa201")
        const aaa202 = graph.getByTestId("aaa202")
        fireEvent.click(aaa201)
        fireEvent.mouseOver(aaa202)
        // not missing because aaa201 is overridden
        expect(aaa101.classList.contains("takeable")).toBe(true)
      })

      it("selected and hovered over: sets itself and all unmet prereqs to be 'missing'", async () => {
        const graph = await TestGraph.build()
        const aaa101 = graph.getByTestId("aaa101")
        const aaa201 = graph.getByTestId("aaa201")
        const aaa102 = graph.getByTestId("aaa102")
        const aaa303 = graph.getByTestId("aaa303")
        fireEvent.click(aaa303)

        expect(aaa101.classList.contains("takeable")).toBe(true)
        expect(aaa102.classList.contains("takeable")).toBe(true)
        expect(aaa201.classList.contains("inactive")).toBe(true)
        expect(aaa303.classList.contains("overridden")).toBe(true)

        fireEvent.mouseOver(aaa303)

        expect(aaa101.classList.contains("missing")).toBe(true)
        expect(aaa102.classList.contains("missing")).toBe(true)
        expect(aaa201.classList.contains("missing")).toBe(true)
        expect(aaa303.classList.contains("missing")).toBe(true)
      })

      it("selected + hovered over + unmet prereqs: 'missing'", async () => {
        const graph = await TestGraph.build()
        const aaa303 = graph.getByTestId("aaa303")
        fireEvent.click(aaa303)
        expect(aaa303.classList.contains("overridden")).toBe(true)

        fireEvent.mouseOver(aaa303)
        expect(aaa303.classList.contains("missing")).toBe(true)
      })
    })
  })
})
