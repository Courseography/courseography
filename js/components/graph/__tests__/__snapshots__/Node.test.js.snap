// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Course Node should match shallow snapshot 1`] = `
<g
  className="node takeable"
  data-testid="csc108"
  id="csc108"
  onClick={[MockFunction]}
  onMouseEnter={[MockFunction]}
  onMouseLeave={[MockFunction]}
  shapeRendering="geometricPrecision"
  textRendering="geometricPrecision"
>
  <rect
    filter="url(#undefined)"
    height={30}
    rx="8"
    ry="8"
    style={
      Object {
        "fill": "#5dd5b8",
      }
    }
    x={594.294128}
    y={38.09390799999999}
  />
  <text
    key="0"
    x={NaN}
    y={59.379107999999995}
  >
    CSC108
  </text>
</g>
`;

exports[`Hybrid Node should match snapshot 1`] = `
<g
  className="hybrid inactive"
  data-testid="h(csc301,csc318,csc384,csc418)"
  id="h46"
  onKeyDown={[MockFunction]}
  shapeRendering="geometricPrecision"
  textRendering="geometricPrecision"
>
  <rect
    filter=""
    height={24}
    style={
      Object {
        "fill": "#888888",
      }
    }
    width={65.207497}
    x={18.168848}
    y={471.586698}
  />
  <text
    key="0"
    x={50.772596500000006}
    y={481.389408}
  >
    CSC318/418/
  </text>
  <text
    key="1"
    x={50.772596500000006}
    y={493.640408}
  >
    301/384
  </text>
</g>
`;
